lib/iris/cube.py

```diff
@@ -2302,7 +2306,8 @@ def _as_list_of_coords(self, names_or_coords):
         Convert a name, coord, or list of names/coords to a list of coords.
         """
         # If not iterable, convert to list of a single item
-        if not hasattr(names_or_coords, '__iter__'):
+        if (not hasattr(names_or_coords, '__iter__') or
+                isinstance(names_or_coords, str)):
```

Comment:
I agree with the new logic in these "iterable but not a string" cases -- i.e. it works and does what we want.
But I think the _reason_ it works is a bit obscure, so I'd prefer :
```
    def is_iterable_but_not_string(x):
        if isinstance(names_or_coords, six.string_types):
            result = False
        else:
            result = isinstance(x, collections.Iterable)
        return result
```

My reasoning is that this test _really_ does just the same thing for either Python2 or 3.
Whereas, the proposed code relies on the slightly odd fact that strings _*don't*_ provide an '\_\_iter\_\_' in Python2 (even though you _*can*_ use "iter(string)", and even though they do have an \_\_iter\_\_ in Python3) :  Without that quirk, this would not work for a unicode object, because they aren't a 'str'.

So, I'm suggesting we should really check for an instance of "basestring" _first_ (equivalently, "six.string_types" for portability).
However, it then also occurred to me that "hasattr(x, '\_\_iter\_\_')"  is not a terribly good test, as objects don't have to provide an "__iter__" method to be iterable (as str+unicode themselves show) :  In fact, you can use "iter(obj)" if obj just has a \_\_getitem\_\_ -- though frustratingly, that does _not_ make it pass the isinstance(obj, collections.Iterable) test !!

Anyway, I believe the above form is also consistent with our usage elsewhere.
e.g. - https://github.com/QuLogic/iris/blob/py3k-unicode/lib/iris/fileformats/cf.py#L1077